src/main/scala/com/databricks/labs/overwatch/utils/Upgrade.scala

```diff
@@ -266,4 +274,570 @@ object Upgrade extends SparkSessionWrapper {
     }
   }
 
+  private def verifyUpgradeStatus(
+                                   upgradeStatus: Array[UpgradeReport],
+                                   initialTableVersions: Map[String, Long],
+                                   snapDir: String): Unit = {
+    if (upgradeStatus.exists(_.failUpgrade)) { // if upgrade report contains fail upgrade
+      val reportTime = System.currentTimeMillis()
+      val upgradeReportPath = s"$snapDir/upgradeReport_$reportTime"
+      val initialTableVersionsPath = s"$snapDir/initialTableVersions_$reportTime"
+
+      val failMsg = s"UPGRADE FAILED:\nUpgrade Report saved as Dataframe to $upgradeReportPath\nTable Versions " +
+        s"prior to upgrade stored as Dataframe at $initialTableVersionsPath"
+
+      logger.log(Level.INFO, failMsg)
+      println(failMsg)
+
+      upgradeStatus.toSeq.toDF().write.format("delta").mode("overwrite")
+        .save(upgradeReportPath)
+
+      initialTableVersions.toSeq.toDF().write.format("delta").mode("overwrite")
+        .save(initialTableVersionsPath)
```

Comment:
will this work in case if we already modified the schema of the table(s)? Maybe we should use `overwriteSchema` ?

Reply:
there's only one write to this path per upgrade init. It's suffixed with the timestamp to the millisecond. This means that if someone needed to resume later they could and then could reference these values each time they kicked off the upgrade.